bee-crypto/src/ternary/sponge/batched_curlp/mod.rs

```diff
@@ -0,0 +1,197 @@
+    /// `N` of each integer.
+    ///
+    /// This step works correctly even if there are less than `BATCH_SIZE` inputs.
+    fn mux(&mut self) {
+        let count = self.trit_inputs.len();
+        for i in 0..self.bct_inputs.len() {
+            let bc_trit = self.bct_inputs.get_mut(i);
+
+            for j in 0..count {
+                match self.trit_inputs[j][i] {
+                    Btrit::NegOne => *bc_trit.lo |= 1 << j,
+                    Btrit::PlusOne => *bc_trit.hi |= 1 << j,
+                    Btrit::Zero => {
+                        *bc_trit.lo |= 1 << j;
+                        *bc_trit.hi |= 1 << j;
+                    }
+                }
+            }
+        }
+    }
+    /// Demultiplex the bits of the output to obtain the hash of the input with an specific index.
+    ///
+    /// This is the inverse of the `mux` function, but it is applied over the vector with the
+    /// binary encoding of the output hashes. Each pair of low and high bits in the `bct_hashes`
+    /// field is decoded into a trit using the same convention as the `mux` step with an additional
+    /// rule for the `(0, 0)` pair of bits which is mapped to the `0` trit.
+    fn demux(&self, index: usize) -> TritBuf {
+        let length = self.bct_hashes.len();
+        let mut result = Vec::with_capacity(length);
+
+        for i in 0..length {
+            let low = (self.bct_hashes.lo()[i] >> index) & 1;
+            // This is safe as the previous access to `self.lo` took care of checking the index.
+            let hi = (unsafe { *self.bct_hashes.hi().get_unchecked(i) } >> index) & 1;
+
+            let trit = match (low, hi) {
+                (1, 0) => Btrit::NegOne,
+                (0, 1) => Btrit::PlusOne,
+                (1, 1) => Btrit::Zero,
+                // This can only be `(0, 0)`.
+                _ => Btrit::Zero,
+            };
+
+            result.push(trit);
+        }
+
+        TritBuf::from_trits(&result)
+    }
+    /// Hash the received inputs using the batched version of CurlP.
+    ///
+    /// This function also takes care of cleaning all the buffers of the struct and resetting the
+    /// batched CurlP hasher so it can be called at any time.
+    pub fn hash_batched<'a>(&'a mut self) -> impl Iterator<Item = TritBuf> + 'a {
+        let total = self.trit_inputs.len();
+        // Fill the `hashes` buffer with zeros.
+        self.bct_hashes.fill(0);
+        // Reset batched CurlP hasher.
+        self.bct_curlp.reset();
+        // Multiplex the trits in `trits` and dump them into `inputs`
+        self.mux();
+        // Do the regular sponge steps.
+        self.bct_curlp.absorb(&self.bct_inputs);
+        self.bct_curlp.squeeze_into(&mut self.bct_hashes);
+        // Clear the `trits` buffer to allow receiving a new batch.
+        self.trit_inputs.clear();
+        // Fill the `inputs` buffer with zeros.
+        self.bct_inputs.fill(0);
```

Comment:
Not sure it's necessary as well ?

Reply:
It is because `mux` doesn't overwrite all the bits. It assumes that before running all the bits are set to zero. But we can fix that.

Reply:
This change actually makes the whole process slower and I don't even understand why. I assume LLVM is particularly good at optimizing the `fill` method.

Reply:
Alright, let keep it as is :)